include/geos/geom/CircularArc.h

```diff
@@ -0,0 +1,268 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2024 ISciences, LLC
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ *
+ **********************************************************************/
+
+#pragma once
+
+#include <geos/export.h>
+#include <geos/geom/Coordinate.h>
+#include <geos/geom/Quadrant.h>
+#include <geos/algorithm/CircularArcs.h>
+#include <geos/algorithm/Orientation.h>
+#include <geos/triangulate/quadedge/TrianglePredicate.h>
+
+namespace geos {
+namespace geom {
+
+/// A CircularArc is a reference to three points that define a circular arc.
+/// It provides for the lazy calculation of various arc properties such as the center, radius, and orientation
+class GEOS_DLL CircularArc {
+public:
+
+    using CoordinateXY = geom::CoordinateXY;
+
+    CircularArc(const CoordinateXY& q0, const CoordinateXY& q1, const CoordinateXY& q2)
+        : p0(q0)
+        , p1(q1)
+        , p2(q2)
+        , m_center_known(false)
+        , m_radius_known(false)
+        , m_orientation_known(false)
+    {}
+
+    const CoordinateXY& p0;
+    const CoordinateXY& p1;
+    const CoordinateXY& p2;
+
+    /// Return the orientation of the arc as one of:
+    /// - algorithm::Orientation::CLOCKWISE,
+    /// - algorithm::Orientation::COUNTERCLOCKWISE
+    /// - algorithm::Orientation::COLLINEAR
+    int orientation() const {
+        if (!m_orientation_known) {
+            m_orientation = algorithm::Orientation::index(p0, p1, p2);
+            m_orientation_known = true;
+        }
+        return m_orientation;
+    }
+
+    /// Return the center point of the circle associated with this arc
+    const CoordinateXY& getCenter() const {
+        if (!m_center_known) {
+            m_center = algorithm::CircularArcs::getCenter(p0, p1, p2);
+            m_center_known = true;
+        }
+
+        return m_center;
+    }
+
+    /// Return the radius of the circle associated with this arc
+    double getRadius() const {
+        if (!m_radius_known) {
+            m_radius = getCenter().distance(p0);
+            m_radius_known = true;
+        }
+
+        return m_radius;
+    }
+
+    /// Return whether this arc forms a complete circle
+    bool isCircle() const {
+        return p0.equals(p2);
+    }
+
+    /// Returns whether this arc forms a straight line (p0, p1, and p2 are collinear)
+    bool isLinear() const {
+        return std::isnan(getRadius());
+    }
+
+    /// Return the inner angle of the sector associated with this arc
+    double getAngle() const {
+        if (isCircle()) {
+            return 2*MATH_PI;
+        }
+
+        auto t0 = theta0();
```

Comment:
potential optimization?: using crossproduct(p0 - center, p2 - center) = radius * radius * sin(angle)  could yield the result by just doing a single asin(), instead of 2 atan2()

Reply:
actually one should also likely compute dotproduct(p0 - center, p2 - center) = radius * radius * cos(angle),  and thus angle = atan2(crossproduct(p0 - center, p2 - center) , dotproduct(p0 - center, p2 - center) )

Reply:
Added as comments